File: app/configs.py
from flask import Blueprint, Response, flash, redirect, render_template, request, url_for

from app.auth import admin_required
from app.config_parser import extract_config_lines
from app.db import (
    delete_configs_by_raw,
    hard_delete_configs,
    import_configs,
    list_config_export_rows,
    list_configs,
)
from app.security import validate_csrf_token

configs_bp = Blueprint("configs", __name__)


@configs_bp.get("/admin/configs")
@admin_required
def config_index():
    configs = list_configs()
    return render_template("configs.html", configs=configs)


@configs_bp.post("/admin/configs/import")
@admin_required
def import_config_batch():
    validate_csrf_token(request.form.get("csrf_token", ""))

    config_blob = request.form.get("config_blob", "")
    config_lines = extract_config_lines(config_blob)

    if not config_lines:
        flash("No valid config lines were found in the pasted text.", "error")
        return redirect(url_for("configs.config_index"))

    stats = import_configs(config_lines)
    flash(
        f"Import complete. Added {stats['inserted']}, skipped {stats['duplicates']} duplicates.",
        "success",
    )
    return redirect(url_for("configs.config_index"))


@configs_bp.post("/admin/configs/delete-by-paste")
@admin_required
def delete_config_batch_by_paste():
    validate_csrf_token(request.form.get("csrf_token", ""))

    config_blob = request.form.get("delete_blob", "")
    config_lines = extract_config_lines(config_blob, exclude_web_urls=False)

    if not config_lines:
        flash("No valid config lines were found in the pasted text.", "error")
        return redirect(url_for("configs.config_index"))

    deleted = delete_configs_by_raw(config_lines)
    not_found = len(config_lines) - deleted
    flash(
        f"Delete complete. Removed {deleted}, skipped {not_found} not found or already deleted.",
        "success",
    )
    return redirect(url_for("configs.config_index"))


@configs_bp.post("/admin/configs/delete")
@admin_required
def batch_delete_configs():
    validate_csrf_token(request.form.get("csrf_token", ""))

    config_ids = parse_config_ids(request.form.getlist("config_ids"))
    if not config_ids:
        flash("Select at least one config to delete.", "error")
        return redirect(url_for("configs.config_index"))

    deleted_count = hard_delete_configs(config_ids)
    flash(f"Deleted {deleted_count} config(s).", "success")
    return redirect(url_for("configs.config_index"))


@configs_bp.post("/admin/configs/<int:config_id>/delete")
@admin_required
def delete_single_config(config_id):
    validate_csrf_token(request.form.get("csrf_token", ""))

    deleted_count = hard_delete_configs([config_id])
    flash(f"Deleted {deleted_count} config(s).", "success")
    return redirect(url_for("configs.config_index"))


@configs_bp.get("/admin/configs/export/all")
@admin_required
def export_all_configs():
    return build_export_response(filename="configs-all.txt")


@configs_bp.get("/admin/configs/export/enabled")
@admin_required
def export_enabled_configs():
    return build_export_response(filename="configs-enabled.txt")


def parse_config_ids(raw_ids):
    parsed = []
    for raw_id in raw_ids:
        try:
            parsed.append(int(raw_id))
        except (TypeError, ValueError):
            continue
    return parsed


def build_export_response(filename):
    rows = list_config_export_rows()
    body = "\n".join(row["raw_config"] for row in rows)
    headers = {
        "Content-Disposition": f'attachment; filename="{filename}"',
    }
    return Response(body, mimetype="text/plain", headers=headers)

File: worker/src/routes/configs.js
import { register } from "../router.js";
import { html, redirect, text } from "../http.js";
import { pushFlash, validateCsrf } from "../context.js";
import { extractConfigLines } from "../parser.js";
import {
  deleteConfigsByRaw,
  hardDeleteConfigs,
  importConfigs,
  listConfigExportRows,
  listConfigs,
} from "../db/index.js";
import { renderConfigsPage } from "../views/configs.js";
import { parseConfigIds, parseIdParam, withAdmin } from "./helpers.js";

register(
  "GET",
  "/admin/configs",
  withAdmin(async (ctx) => html(renderConfigsPage(ctx, await listConfigs(ctx.env)))),
);

register(
  "POST",
  "/admin/configs/import",
  withAdmin(async (ctx) => {
    const form = await ctx.request.formData();
    validateCsrf(ctx, form.get("csrf_token") || "");
    const lines = extractConfigLines(form.get("config_blob") || "");
    if (!lines.length) {
      pushFlash(ctx, "error", "No valid config lines were found in the pasted text.");
      return redirect("/admin/configs");
    }
    const stats = await importConfigs(ctx.env, lines);
    pushFlash(
      ctx,
      "success",
      `Import complete. Added ${stats.inserted}, skipped ${stats.duplicates} duplicates.`,
    );
    return redirect("/admin/configs");
  }),
);

register(
  "POST",
  "/admin/configs/delete-by-paste",
  withAdmin(async (ctx) => {
    const form = await ctx.request.formData();
    validateCsrf(ctx, form.get("csrf_token") || "");
    const lines = extractConfigLines(form.get("delete_blob") || "", { excludeWebUrls: false });
    if (!lines.length) {
      pushFlash(ctx, "error", "No valid config lines were found in the pasted text.");
      return redirect("/admin/configs");
    }
    const deleted = await deleteConfigsByRaw(ctx.env, lines);
    const notFound = lines.length - deleted;
    pushFlash(
      ctx,
      "success",
      `Delete complete. Removed ${deleted}, skipped ${notFound} not found or already deleted.`,
    );
    return redirect("/admin/configs");
  }),
);

register(
  "POST",
  "/admin/configs/delete",
  withAdmin(async (ctx) => {
    const form = await ctx.request.formData();
    validateCsrf(ctx, form.get("csrf_token") || "");
    const configIds = parseConfigIds(form.getAll("config_ids"));
    if (!configIds.length) {
      pushFlash(ctx, "error", "Select at least one config to delete.");
      return redirect("/admin/configs");
    }
    const deletedCount = await hardDeleteConfigs(ctx.env, configIds);
    pushFlash(ctx, "success", `Deleted ${deletedCount} config(s).`);
    return redirect("/admin/configs");
  }),
);

register(
  "POST",
  "/admin/configs/:id/delete",
  withAdmin(async (ctx, params) => {
    const form = await ctx.request.formData();
    validateCsrf(ctx, form.get("csrf_token") || "");
    const deletedCount = await hardDeleteConfigs(ctx.env, [parseIdParam(params.id)]);
    pushFlash(ctx, "success", `Deleted ${deletedCount} config(s).`);
    return redirect("/admin/configs");
  }),
);

register(
  "GET",
  "/admin/configs/export/all",
  withAdmin((ctx) => buildExportResponse(ctx.env, "configs-all.txt")),
);

register(
  "GET",
  "/admin/configs/export/enabled",
  withAdmin((ctx) => buildExportResponse(ctx.env, "configs-enabled.txt")),
);

async function buildExportResponse(env, filename) {
  const rows = await listConfigExportRows(env);
  const body = rows.map((row) => row.raw_config).join("\n");
  return {
    ...text(body),
    headers: { "Content-Disposition": `attachment; filename="${filename}"` },
  };
}
